### PR TITLE
Campaign: New loadout system and perk system

### PR DESCRIPTION
**Pull request by Lumipharon**
## About The Pull Request
Requires/includes #14776 so probs review/merge that first.

Ready for review. Theres a few placeholder icons and some of the prices are just fairly random but I can sort those out down the line.

### Loadouts
Adds a new loadout system for Campaign. You now get (some) customisation back.
image

A new action button lets you open up a UI where you can customise your loadout for all available jobs in your faction.
Every slot has a list of selectable items.
You populate all your choices, then hit equip and you're good to go.

https://github.com/tgstation/TerraGov-Marine-Corps/assets/7869430/911a4157-480e-48fc-a9c8-699e1c918119

The loadout(s) are preserved for your character, so you can adjust or change them next mission, or while you're dead etc.

Some items either having a whitelist (i.e. you need another item to equip it) or a blacklist (equipping this forbids other items).
Some can have an equip cost (i.e. a cost every time you equip a loadout with it), or have to be unlocked with a one off cost.
Certain selections (mainly your main weapon) also help determine the contents of your loadout. i.e. guns with a UGL will likely have grenades in their bag or webbing, etc).

### Perks
You can now unlock perks that apply permanent bonuses to your character.
image

These vary from job to job, but most are universal. i.e. most gun skills.
Most perks are skill related currently, but there are others, such as max HP boosts, the ability to walk silently, and a couple to unlock special melee abilities etc.

Loadouts are still mainly confined to existing available stuff in campaign (SOM especially don't really have any cosmetic items since none exist for them), but this will expand in the future after this is successfully released.
Likewise I have plans to greatly expand the perk system.

I also hope to make the interface look a lot nicer, someone please pay gregor so I can have big icon buttons instead of what I currently have.
## Why It's Good For The Game
Customisation without 20 minute prep time is good. Individual progression equals far better player investment.
## Changelog
:cl:
add: Added a new perk system for Campaign mode
add: Added a new customisable loadout system for Campaign mode
/:cl:

### CODE REVIEW

**Comment by Loafin34** [comment]
Will one of the perks be mag switching? (tap gun with mag to reload it without caring if mag was out of gun or not)

**Review by Kuro020 — APPROVED** [review]

**Comment by Lumipharon** [comment]
> Will one of the perks be mag switching? (tap gun with mag to reload it without caring if mag was out of gun or not)

Uh, you mean tactical reloads? That's already a thing.

**Comment by shelteredfreak** [comment]
>adds Perks
Can we play Campaign 24/7 now? Please and thank you. 

I might miss xenos, but maybe you can add some xeno missions and xeno support to Campaign someday...

File: code/__DEFINES/assert.dm [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#undef ASSERT
+
+/// Override BYOND's native ASSERT to optionally specify a message
+#define ASSERT(condition, message...) \
+	if (!(condition)) { \
+		CRASH(assertion_message(__FILE__, __LINE__, #condition, ##message)) \
+	}
+
+/proc/assertion_message(file, line, condition, message)
```
Comment (ivanmixo) [inline code/__DEFINES/assert.dm:9]: doc

File: code/datums/gamemodes/campaign/individual_stats.dm [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+	if(length(new_perk.prereq_perks))
+		var/perk_found
+		for(var/prereq in new_perk.prereq_perks)
+			perk_found = FALSE
+			for(var/datum/perk/perk AS in unlocked_perks)
+				if(perk.type != prereq)
+					continue
+				perk_found = TRUE
+			if(!perk_found)
+				to_chat(user, "<span class='warning'>One or more prerequisites missing for this perk.")
```
Comment (ivanmixo) [inline code/datums/gamemodes/campaign/individual_stats.dm:70]: use the macros for these spans

**Review by ivanmixo — CHANGES_REQUESTED** [review]
Minor stuff

**Review by ivanmixo — APPROVED** [review]

**Comment by Tommy-211** [comment]
You know what i love? more hvh game modes that nobody will play after the first week of it's release

**Comment by chizzy376** [comment]
I like this approach, but I wish it told you what the skills do when upgraded. I.e., corpsman getting the better medicine perk, how much they benefit from it

**Pull request merged**